Added timedelta feature and timezone conversion

bakayu opened the pull request:
closes #1 closes #2 closes #3

bakayu committed:
Added timedelta feature and timezone conversion

dubesar commented:
Hi @bakayu, please update the testing done as well in the details.

sPredictorX1708 commented on worldtimebuddy/cli.py:
```diff
@@ -21,9 +37,27 @@ def cli(format, major, timezone, list_timezones):
         try:
             tz = pytz.timezone(timezone)
             time = now.astimezone(tz)
+            
+            if delta:
+                if 'hr' in delta:
+                    hours = int(delta.replace('hr', ''))
+                    time += datetime.timedelta(hours=hours)
+                elif 'min' in delta:
+                    minutes = int(delta.replace('min', ''))
+                    time += datetime.timedelta(minutes=minutes)
+                elif 'day' in delta:
+                    days = int(delta.replace('day', ''))
+                    time += datetime.timedelta(days=days)
+                else:
+                    click.echo("Error: Invalid time format. Use +2hr, -30min, -1day, etc.")
+                    return
+
```
Consider this to add in utils

bakayu replied:
can you kindly elaboraite on this? thank you.

sPredictorX1708 replied:
Consider moving the block 

```
if 'hr' in delta:
    hours = int(delta.replace('hr', ''))
    time += datetime.timedelta(hours=hours)
elif 'min' in delta:
    minutes = int(delta.replace('min', ''))
    time += datetime.timedelta(minutes=minutes)
elif 'day' in delta:
    days = int(delta.replace('day', ''))
    time += datetime.timedelta(days=days)
```
into some util function in utils.py file.

sPredictorX1708 commented on worldtimebuddy/cli.py:
```diff
@@ -4,12 +4,28 @@
 MAJOR_TIMEZONES = ['UTC', 'US/Pacific', 'Asia/Kolkata']
 
+# Mapping of short timezone codes to full names
+timezone_codes = {
+    'IST': 'Asia/Kolkata',
+    'PST': 'US/Pacific',
+    'EST': 'US/Eastern',
+    'CST': 'US/Central',
+    'MST': 'US/Mountain',
+    'GMT': 'Europe/London',
+    'CET': 'Europe/Paris',
+    'EET': 'Europe/Athens',
+    'JST': 'Asia/Tokyo',
+    'AEST': 'Australia/Sydney',
+    'UTC': 'UTC'
+}
```
Consider this adding a separate constants file.

bakayu replied:
will get back to it right away, I worked on it before you added the utils.py file :)

sPredictorX1708 replied:
Also please add testing details for both the changes.

bakayu replied:
> Also please add testing details for both the changes.

would screen shots be enough?

sPredictorX1708 replied:
Can you check the PR - https://github.com/dubesar/worldtimebuddy/pull/7, attach similarly with different cases.

bakayu committed:
Merge branch 'main' of https://github.com/dubesar/worldtimebuddy

bakayu committed:
modified cli.py utils.py and added constants.py
added the delta function to the cli, moved extra functions to utils.py
added timezone conversion codes feature, see constants.py file for supported timezones

bakayu commented:
testing:

timedelta and usage of short timezone codes
``` bash
worldtimebuddy --timezone IST --delta +2hr
Asia/Kolkata                   2024-10-07 01:40:07
```
``` bash
worldtimebuddy --timezone JST --delta +5day
Asia/Tokyo                     2024-10-12 03:10:11
```
``` bash
worldtimebuddy --timezone UTC --delta -1min
UTC                            2024-10-06 18:09:13
```

![image](https://github.com/user-attachments/assets/159e2975-1a47-4881-b6d5-67fe7ba71588)

sPredictorX1708 commented on worldtimebuddy/cli.py:
```diff
@@ -11,7 +12,8 @@
 @click.option('--timezone', '-tz', help='Show time for a specific timezone')
 @click.option('--list', 'list_timezones', is_flag=True, help='List all available timezones')
-def cli(format, major, timezone, list_timezones):
+@click.option('--delta', help='Time to add (e.g., +2hr, -30min, -1day)')
```
Please modify readme for new parameter and its usage.

bakayu replied:
on it

sPredictorX1708 commented on worldtimebuddy/utils.py:
```diff
@@ -1,9 +1,24 @@
+        days = int(delta.replace('day', ''))
+        time += datetime.timedelta(days=days)
+    else:
+        raise ValueError("Invalid time format. Use +2hr, -30min, -1day, etc.")
```
If returning a datetime how will the ValueError behave? Can you think of better approach here for return type? Also please attach testing for negative case as well where explicitly the else case gets executed.

bakayu replied:
> If returning a datetime how will the ValueError behave? Can you think of better approach here for return type? 

should I define a custom error in case an invalid argument is passed and else case is executed?

sPredictorX1708 replied:
I think ValueError seems fine, but in that case there is no return value right? How are we handling that?

bakayu replied:
> If returning a datetime how will the ValueError behave? Can you think of better approach here for return type?

when ValueError is raised its caught by the try catch block in cli.py which calls the delta function
``` py
if delta:
    try:
         time = callDelta(time, delta)
    except ValueError as e:
         click.echo(f"Error: {e}")
         return
```

I can rename it to a custom error to avoid confusion, will that do?

sPredictorX1708 replied:
Yes, also update the testing for that.

bakayu committed:
updated README.md, added custom error for delta function
updated README.md to include installation guide and usage of delta function,
defined a custom error `DeltaValueError` for when wrong arguments are passed through delta function

bakayu commented:
Updated as per your feedback :)

additional testing:
```
worldtimebuddy --timezone UTC --delta +2
DeltaValueError: Invalid time format. Use +2hr, -30min, -1day, etc.
```
```
worldtimebuddy --timezone IST --delta -1day
Asia/Kolkata                   2024-10-06 00:26:31
```
```
worldtimebuddy --timezone IST --delta -1hr
Asia/Kolkata                   2024-10-06 23:26:35
```

![image](https://github.com/user-attachments/assets/caf5a305-4ad8-43f5-b228-4bacafc39d8f)

bakayu commented:
I have updated the README.md file to include delta function use cases and added a insallation guide as well

dubesar commented:
LGTM, will approve and merge after one more round of review!

sPredictorX1708 commented on README.md:
````diff
@@ -16,8 +16,22 @@ Options:
+2. Run the following code to install dependencies and worldtimebuddy
+``` bash
+pip install -r requirements.txt
+python3 setup.py built
````
This should be `python3 setup.py build`

bakayu replied:
my bad, fixing the typo

sPredictorX1708 commented on README.md:
```diff
@@ -75,4 +89,20 @@ Africa/Algiers
+
+
+6. Getting a difference of time from a particular timezone 
+(supported time formats: +/- min, hr, day)
```
Put a line break here.

bakayu replied:
the whole usage section is in a single code block, should I fix it and put individual sample code in code blocks?

sPredictorX1708 replied:
No just put an extra line after line#96

sPredictorX1708 replied:
> the whole usage section is in a single code block, should I fix it and put individual sample code in code blocks?

This would be great! If you can do it?

sPredictorX1708 commented on README.md:
```diff
@@ -75,4 +89,20 @@ Africa/Algiers
+
+UTC                            2024-10-06 18:43:35
+
+
```
Remove one empty line from here.

bakayu committed:
updated README.md

bakayu commented:
updated README.md file

bakayu committed:
Merge branch 'main' into timedelta-timeconversion

sPredictorX1708 commented on README.md:
````diff
@@ -16,35 +16,53 @@ Options:
 versions.
+eg.
+```
+> export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
+> worldtimebuddy --major
 
 eg. export MAJOR_TIMEZONES=["UTC","Asia/Kolkata"]
 
````
Seems duplicate

sPredictorX1708 replied:
Btw, Line#67 is correct one, Line#64 won't work.

sPredictorX1708 commented on README.md:
````diff
@@ -57,22 +75,39 @@ Asia/Kolkata                   2024-10-06 19:20:26
+```
+
+6. Getting a difference of time from a particular timezone <br> (supported time formats: +/- min, hr, day) 
+`without --delta argument`
+```
+> worldtimebuddy --timezone UTC
````
Is this example required? Can we just show an extra example in point 5? Like

```
> worldtimebuddy --timezone JST
```

??

bakayu replied:
done, please review

bakayu committed:
updated README.md

bakayu committed:
Merge branch 'timedelta-timeconversion' of https://github.com/bakayu/worldtimebuddy into timedelta-timeconversion

bakayu committed:
update README.md

bakayu commented:
please review.

sPredictorX1708 commented on README.md:
```diff
@@ -64,17 +64,9 @@ eg.
 > export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
```
This should be MAJOR_TIMEZONES=["UTC","Asia/Kolkata"], Please make this small change, rest LGTM. Thanks

bakayu replied:
done

bakayu committed:
updated README.md

sPredictorX1708 commented on README.md:
````diff
@@ -61,7 +61,7 @@ Currenlty only IST/PST/PDT/UTC by default. But user can set major timezones in t
 versions.
 eg.
 ```
-> export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
````
I meant 

export MAJOR_TIMEZONES=["UTC","Asia/Kolkata"]
Please check the comment above, we cannot have space in the array.

bakayu replied:
I see, updated with appropriate changes.

bakayu committed:
updated README.md

sPredictorX1708 reviewed (approved):
LGTM, Thanks for the change!

dubesar commented:
Merged!

bakayu commented:
please add the hacktoberfest-accepted label, thanks! ^^
